javascript/ql/src/semmle/javascript/StringOps.qll

```diff
@@ -0,0 +1,398 @@
+      result = getArgument(0)
+    }
+
+    override DataFlow::Node getSubstring() {
+      result = getArgument(1)
+    }
+  }
+
+  /**
+   * A check of form `A.indexOf(B) !== -1` or similar.
+   */
+  private class Includes_IndexOfEquals extends Includes, DataFlow::ValueNode {
+    MethodCallExpr indexOf;
+    override EqualityTest astNode;
+
+    Includes_IndexOfEquals() {
+      exists (Expr index | astNode.hasOperands(indexOf, index) |
+        // one operand is of the form `whitelist.indexOf(x)`
+        indexOf.getMethodName() = "indexOf" and
+        // and the other one is -1
+        index.getIntValue() = -1
+      )
+    }
+
+    override DataFlow::Node getBaseString() {
+      result = indexOf.getReceiver().flow()
+    }
+
+    override DataFlow::Node getSubstring() {
+      result = indexOf.getArgument(0).flow()
+    }
+
+    override boolean getPolarity() {
+      result = astNode.getPolarity().booleanNot()
+    }
+  }
+
+  /**
+   * A check of form `A.indexOf(B) >= 0` or similar.
+   */
+  private class Includes_IndexOfRelational extends Includes, DataFlow::ValueNode {
+    MethodCallExpr indexOf;
+    override RelationalComparison astNode;
+    boolean polarity;
+
+    Includes_IndexOfRelational() {
+      exists (Expr lesser, Expr greater |
+        astNode.getLesserOperand() = lesser and
+        astNode.getGreaterOperand() = greater and
+        indexOf.getMethodName() = "indexOf" and
+        indexOf.getNumArgument() = 1 |
+        polarity = true and
+        greater = indexOf and
+        (
+          lesser.getIntValue() = 0 and astNode.isInclusive()
+          or
+          lesser.getIntValue() = -1 and not astNode.isInclusive()
+        )
+        or
+        polarity = false and
+        lesser = indexOf and
+        (
+          greater.getIntValue() = -1 and astNode.isInclusive()
+          or
+          greater.getIntValue() = 0 and not astNode.isInclusive()
+        )
+      )
+    }
+
+    override DataFlow::Node getBaseString() {
+      result = indexOf.getReceiver().flow()
+    }
+
+    override DataFlow::Node getSubstring() {
+      result = indexOf.getArgument(0).flow()
+    }
+
+    override boolean getPolarity() {
+      result = polarity
+    }
+  }
+
+  /**
+   * An expression of form `~A.indexOf(B)` which, when coerced to a boolean, is equivalent to `A.includes(B)`.
+   */
+  private class Includes_IndexOfBitwise extends Includes, DataFlow::ValueNode {
+    MethodCallExpr indexOf;
+    override BitNotExpr astNode;
+
+    Includes_IndexOfBitwise() {
+      astNode.getOperand() = indexOf and
+      indexOf.getMethodName() = "indexOf"
+    }
+
+    override DataFlow::Node getBaseString() {
+      result = indexOf.getReceiver().flow()
+    }
+
+    override DataFlow::Node getSubstring() {
+      result = indexOf.getArgument(0).flow()
+    }
+  }
+
+  /**
+   * An expression that appears to be part of an `endsWith`-check, that is,
+   * roughly equivalent to `A.endsWith(B)` or `!A.endsWith(B)`.
```

Comment:
What does "roughly" mean here?

Reply:
Ah, good catch. I wrote this before I decided to exclude expressions that aren't strictly equivalent to `endsWith`, such as `A.substr(-B.length) === B`. I've updated the comment.